controller/snapshot_controller.go

```diff
@@ -0,0 +1,545 @@
+}
+
+func (sc *SnapshotController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync snapshot %v", sc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != sc.namespace {
+		return nil
+	}
+	return sc.reconcile(name)
+}
+
+func (sc *SnapshotController) reconcile(snapshotName string) (err error) {
+	snapshot, err := sc.ds.GetSnapshot(snapshotName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	isResponsible, err := sc.isResponsibleFor(snapshot)
+	if err != nil {
+		return err
+	}
+	if !isResponsible {
+		return nil
+	}
+
+	if snapshot.Status.OwnerID != sc.controllerID {
+		snapshot.Status.OwnerID = sc.controllerID
+		snapshot, err = sc.ds.UpdateSnapshotStatus(snapshot)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+	}
+
+	log := getLoggerForSnapshot(sc.logger, snapshot)
+
+	// deleting snapshotCR
+	if !snapshot.DeletionTimestamp.IsZero() {
+		volume, err := sc.ds.GetVolumeRO(snapshot.Spec.Volume)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Infof("Removing finalizer for snapshot %v", snapshot.Name)
+				return sc.ds.RemoveFinalizerForSnapshot(snapshot)
+			}
+			return err
+		}
+		if !volume.DeletionTimestamp.IsZero() {
+			log.Infof("Removing finalizer for snapshot %v", snapshot.Name)
+			return sc.ds.RemoveFinalizerForSnapshot(snapshot)
+		}
+		engine, err := sc.getTheOnlyEngineCRforSnapshot(snapshot)
+		if err != nil {
+			return err
+		}
+		if engine.Status.CurrentState != longhorn.InstanceStateRunning {
+			// TODO: how to prevent duplicated event here
+			sc.eventRecorder.Eventf(snapshot, v1.EventTypeWarning, "SnapshotDeleteError", "cannot delete snapshot because the volume engine %v is not running", engine.Name)
+			sc.logger.Errorf("cannot delete snapshot because the volume engine %v is not running", engine.Name)
+			return nil
+		}
+		// Delete the snapshot from engine process
+		if err := sc.handleSnapshotDeletion(snapshot, engine); err != nil {
+			return err
+		}
+
+		// Wait for the snapshot to be removed from engine.Status.Snapshots
+		engine, err = sc.ds.GetEngineRO(engine.Name)
+		if err != nil {
+			return err
+		}
+		snapshotInfo, ok := engine.Status.Snapshots[snapshot.Name]
+		if !ok {
+			log.Infof("Removing finalizer for snapshot %v", snapshot.Name)
+			return sc.ds.RemoveFinalizerForSnapshot(snapshot)
+		}
+
+		existingSnapshot := snapshot.DeepCopy()
+		if err := syncSnapshotWithSnapshotInfo(snapshot, snapshotInfo, engine.Spec.VolumeSize); err != nil {
+			return err
+		}
+		if reflect.DeepEqual(existingSnapshot.Status, snapshot.Status) {
+			return nil
+		}
+		if _, err = sc.ds.UpdateSnapshotStatus(snapshot); err != nil {
+			return err
+		}
+		sc.generatingEventsForSnapshot(existingSnapshot, snapshot)
+		return nil
```

Comment:
> I guess it's for the snapshots that cannot be purged immediately.

You are correct

I would like to seperate the code path for snapshot with deletion timestamp out of the code path for snapshot creation and updating. Yeah, the code looks longer but it simply the logic in the creation/updating path because we don't have to worry about snapshot is being deleted. 

How do you think?